eZ/Bundle/EzPublishCoreBundle/ApiLoader/SearchEngineFactory.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * This file is part of the eZ Publish Kernel package
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ * @version //autogentag//
+ */
+
+namespace eZ\Bundle\EzPublishCoreBundle\ApiLoader;
+
+use eZ\Bundle\EzPublishCoreBundle\ApiLoader\Exception\InvalidSearchEngine;
+use eZ\Publish\SPI\Search\Handler as SearchHandler;
+
+/**
+ * The search engine factory.
+ */
+class SearchEngineFactory
+{
+    /**
+     * @var \eZ\Bundle\EzPublishCoreBundle\ApiLoader\StorageRepositoryProvider
+     */
+    private $storageRepositoryProvider;
+
+    /**
+     * Hash of registered search engines.
+     * Key is the search engine identifier, value search handler itself.
+     *
+     * @var \eZ\Publish\SPI\Search\Handler[]
+     */
+    protected $searchEngines = array();
+
+    public function __construct( StorageRepositoryProvider $storageRepositoryProvider )
+    {
+        $this->storageRepositoryProvider = $storageRepositoryProvider;
+    }
+
+    /**
+     * Registers $searchHandler as a valid search engine with identifier $searchEngineIdentifier.
+     *
+     * @note It is strongly recommenced to register a lazy persistent handler.
+     *
+     * @param \eZ\Publish\SPI\Search\Handler $searchHandler
+     * @param string $searchEngineIdentifier
+     */
+    public function registerSearchEngine( SearchHandler $searchHandler, $searchEngineIdentifier )
+    {
+        $this->searchEngines[$searchEngineIdentifier] = $searchHandler;
+    }
+
+    /**
+     * Returns registered search engines.
+     *
+     * @return \eZ\Publish\SPI\Search\Handler[]
+     */
+    public function getSearchEngines()
+    {
+        return $this->searchEngines;
+    }
+
+    /**
+     * Builds search engine identified by its identifier (the "alias" attribute in the service tag),
+     * resolved for current siteaccess.
+     *
+     * @throws \eZ\Bundle\EzPublishCoreBundle\ApiLoader\Exception\InvalidSearchEngine
+     *
+     * @return \eZ\Publish\SPI\Persistence\Handler
+     */
+    public function buildSearchEngine()
+    {
+        $repositoryConfig = $this->storageRepositoryProvider->getRepositoryConfig();
+
+        if (
+            !isset( $repositoryConfig['search']['engine'] )
+            || !isset( $this->searchEngines[$repositoryConfig['search']['engine']] )
+        )
+        {
+            throw new InvalidSearchEngine(
+                "Invalid search engine '{$repositoryConfig['search']['engine']}'. " .
```

Comment:
On second look, these exceptions extend `InvalidArgumentException`, which IMHO does not make much sense. There is Symfony's [InvalidConfigurationException](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Config/Definition/Exception/InvalidConfigurationException.php) that should be a better match. So I'm just going to keep `InvalidSearchEngine` and [these two](https://github.com/ezsystems/ezpublish-kernel/tree/master/eZ/Bundle/EzPublishCoreBundle/ApiLoader/Exception) as is and leave it for later. (Issue pending)